File: java/src/hu/klenium/tetris/util/Point.java
package hu.klenium.tetris.util;

import java.util.Objects;

public class Point {
    public final int x;
    public final int y;
    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }
    public Point add(Point other) {
        return new Point(this.x + other.x, this.y + other.y);
    }
    @Override public boolean equals(Object obj) {
        if (obj == this)
            return true;
        if (obj == null || getClass() != obj.getClass())
            return false;
        Point other = (Point) obj;
        return this.x == other.x && this.y == other.y;
    }
    @Override public int hashCode() {
        return Objects.hash(x, y);
    }
}

File: kotlin/src/hu/klenium/tetris/util/Point.kt
package hu.klenium.tetris.util

data class Point(val x: Int, val y: Int) {
    constructor(size: Dimension) : this(size.width, size.height)
    operator fun plus(other: Point): Point {
        return Point(this.x + other.x, this.y + other.y)
    }
    operator fun plus(size: Dimension) : Point {
        return Point(this.x + size.width, this.y + size.height)
    }
    infix fun notSmallerThan(other: Point) : Boolean {
        return this.x >= other.x && this.y >= other.y
    }
    infix fun notGreaterThan(other: Point) : Boolean {
        return this.x <= other.x && this.y <= other.y
    }
}